### PR TITLE
Convert service config code to use new JSON API

### PR DESCRIPTION
**Pull request by markdroth**

### CODE REVIEW

**Comment by markdroth** [comment]
@apolcyn It looks like I'm running into another hard-to-debug resolver_component_test failure here:

https://source.cloud.google.com/results/invocations/ef556a45-526a-4c50-ac00-50433e4a7bfe/targets

Based on the expected service config error message ("Service config parsing error"), it looks like it's coming from this test case:
 https://github.com/grpc/grpc/blob/5791495f595b30545f3dfd31321f4b5760b11c75/test/cpp/naming/resolver_test_record_groups.yaml#L437

However, looking at the data for that case, I don't actually see why the input data would cause a failure.  The only thing strange about the input data is that it has the "waitForReady" field inside of the "name" field, but the parser should just treat that as an unknown field and ignore it.  And indeed, with this PR, it looks like that test case is not causing an error, which looks fine.

So, there are two things I don't understand about this.  First, why was this test case added, and what is it actually trying to test?  And second, why was the test passing prior to this PR, given that this input should never have been causing a parse error in the first place?

I'd appreciate any help you can offer with this.  Thanks!

**Comment by apolcyn** [comment]
> First, why was this test case added, and what is it actually trying to test? And second, why was the test passing prior to this PR, given that this input should never have been causing a parse error in the first place?

This assertion was added in https://github.com/grpc/grpc/pull/18946. cc @yashykt for insight on motivation.

I just ran the test on master and logged the full service config error string that is currently getting generated for that test:

```
D0124 10:26:39.439697244  101098 resolver_component_test.cc:274] got service config error: {"created":"@1579890399.438452847","description":"Service config parsing error","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":82,"referenced_errors":[{"created":"@1579890399.438448266","description":"Method Params","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":213,"referenced_errors":[{"created":"@1579890399.438287347","description":"methodConfig","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":157,"referenced_errors":[{"created":"@1579890399.438252777","description":"field:name error:Child entry not of type string","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":249},{"created":"@1579890399.438284915","description":"No names specified","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":148}]},{"created":"@1579890399.438300548","description":"methodConfig","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":157,"referenced_errors":[{"created":"@1579890399.438297815","description":"field:name error:Child entry not of type string","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":249},{"created":"@1579890399.438299168","description":"No names specified","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":148}]},{"created":"@1579890399.438310828","description":"methodConfig","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":157,"referenced_errors":[{"created":"@1579890399.438308293","description":"field:name error:Child entry not of type string","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":249},{"created":"@1579890399.438309472","description":"No names specified","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":148}]},{"created":"@1579890399.438321620","description":"methodConfig","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":157,"referenced_errors":[{"created":"@1579890399.438318884","description":"field:name error:Child entry not of type string","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":249},{"created":"@1579890399.438320277","description":"No names specified","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":148}]},{"created":"@1579890399.438332201","description":"methodConfig","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":157,"referenced_errors":[{"created":"@1579890399.438329632","description":"field:name error:Child entry not of type string","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":249},{"created":"@1579890399.438330868","description":"No names specified","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":148}]},{"created":"@1579890399.438344229","description":"methodConfig","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":157,"referenced_errors":[{"created":"@1579890399.438341499","description":"field:name error:Child entry not of type string","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":249},{"created":"@1579890399.438342722","description":"No names specified","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":148}]},{"created":"@1579890399.438355338","description":"methodConfig","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":157,"referenced_errors":[{"created":"@1579890399.438352635","description":"field:name error:Child entry not of type string","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":249},{"created":"@1579890399.438353851","description":"No names specified","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":148}]},{"created":"@1579890399.438366063","description":"methodConfig","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":157,"referenced_errors":[{"created":"@1579890399.438363397","description":"field:name error:Child entry not of type string","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":249},{"created":"@1579890399.438364616","description":"No names specified","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":148}]},{"created":"@1579890399.438376687","description":"methodConfig","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":157,"referenced_errors":[{"created":"@1579890399.438374041","description":"field:name error:Child entry not of type string","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":249},{"created":"@1579890399.438375264","description":"No names specified","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":148}]},{"created":"@1579890399.438389317","description":"methodConfig","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":157,"referenced_errors":[{"created":"@1579890399.438386436","description":"field:name error:Child entry not of type string","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":249},{"created":"@1579890399.438387663","description":"No names specified","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":148}]},{"created":"@1579890399.438400265","description":"methodConfig","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":157,"referenced_errors":[{"created":"@1579890399.438397459","description":"field:name error:Child entry not of type string","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":249},{"created":"@1579890399.438398788","description":"No names specified","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":148}]},{"created":"@1579890399.438410696","description":"methodConfig","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":157,"referenced_errors":[{"created":"@1579890399.438408110","description":"field:name error:Child entry not of type string","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":249},{"created":"@1579890399.438409313","description":"No names specified","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":148}]},{"created":"@1579890399.438421477","description":"methodConfig","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":157,"referenced_errors":[{"created":"@1579890399.438418835","description":"field:name error:Child entry not of type string","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":249},{"created":"@1579890399.438420051","description":"No names specified","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":148}]},{"created":"@1579890399.438432236","description":"methodConfig","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":157,"referenced_errors":[{"created":"@1579890399.438429583","description":"field:name error:Child entry not of type string","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":249},{"created":"@1579890399.438430789","description":"No names specified","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":148}]},{"created":"@1579890399.438443021","description":"methodConfig","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":157,"referenced_errors":[{"created":"@1579890399.438440501","description":"field:name error:Child entry not of type string","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":249},{"created":"@1579890399.438441688","description":"No names specified","file":"src/core/ext/filters/client_channel/service_config.cc","file_line":148}]}]}]}

```

with the main error being:

```
"field:name error:Child entry not of type string"
```

**Comment by apolcyn** [comment]
continuing on that, I think the problem with that service config is, like you said, that the "wait for ready" field is inside of the "name" object within the service config.

The current service config validation fails on https://github.com/grpc/grpc/blob/5791495f595b30545f3dfd31321f4b5760b11c75/src/core/ext/filters/client_channel/service_config.cc#L249, because it expects all fields within the name object to be of type string, but the "waitForReady" field being there naturally breaks that.

If this an unknown field like that should be simply ignored, then it would seem like the current service config validation logic has a problem in following that piece of the spec, and that the test also doesn't follow that piece of the spec.

cc @yashykt

**Comment by markdroth** [comment]
Ah!  Thanks for the debugging help -- that tells me exactly what I needed to know.

I'm pretty sure the old code was being overly picky there.  It should not have been checking the type of any unknown fields.  So the new code is correct in that regard.

However, we do still need a test to cover the case where there is an invalid service config, so I'll modify the test to specify a different invalid config.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc [hunk 1]
```diff
@@ -223,100 +223,88 @@ void AresDnsResolver::OnNextResolutionLocked(void* arg, grpc_error* error) {
...
+  if (*error != GRPC_ERROR_NONE) return nullptr;
+  if (json.type() != Json::Type::ARRAY) {
     *error = GRPC_ERROR_CREATE_FROM_STATIC_STRING(
         "Service Config Choices, error: should be of type array");
     return nullptr;
   }
-  char* service_config = nullptr;
+  std::string service_config = nullptr;
   InlinedVector<grpc_error*, 4> error_list;
-  bool found_choice = false;  // have we found a choice?
```
Comment (yashykt) [inline src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc]: looks like there is a semantic change here. Earlier, we were validating all the choices, but now we choose the first choice and exit. The proposal suggests that we should be using the first choice that fits, so the new code seems right.

File: test/cpp/naming/resolver_component_tests_runner.py [hunk 2]
```diff
@@ -555,8 +555,8 @@ def _quit_on_signal(signum, _frame):
...
...bin","methodConfig":[{"name":[{"method":"Foo","service":"SimpleService","waitForReady":true}]},{"name":[{"method":"FooTwo","service":"SimpleService","waitForReady":true}]},{"name":[{"method":"FooThree","service":"SimpleService","waitForReady":true}]},{"name":[{"method":"FooFour","service":"SimpleService","waitForReady":true}]},{"name":[{"method":"FooFive","service":"SimpleService","waitForReady":true}]},{"name":[{"method":"FooSix","service":"SimpleService","waitForReady":true}]},{"name":[{"method":"FooSeven","service":"SimpleService","waitForReady":true}]},{"name":[{"method":"FooEight","service":"SimpleService","waitForReady":true}]},{"name":[{"method":"FooNine","service":"SimpleService","waitForReady":true}]},{"name":[{"method":"FooTen","service":"SimpleService","waitForReady":true}]},{"name":[{"method":"FooEleven","service":"SimpleService","waitForReady":true}]},{"name":[{"method":"FooTwelve","service":"SimpleService","waitForReady":true}]},{"name":[{"method":"FooTwelve","service":"SimpleService","waitForReady":true}]},{"name":[{"method":"FooTwelve","service":"SimpleService","waitForReady":true}]},{"name":[{"method":"FooTwelve","service":"SimpleService","waitForReady":true}]}]}',
```
Comment (yashykt) [inline test/cpp/naming/resolver_component_tests_runner.py]: what was the reason to remove this?

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc (hunk 1 above)
Comment (markdroth) [reply src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc]: Hmm, yeah, good point.  We probably should actually validate all of the options.  I've changed the code to preserve the old behavior.

File: test/cpp/naming/resolver_component_tests_runner.py (hunk 2 above)
Comment (markdroth) [reply test/cpp/naming/resolver_component_tests_runner.py]: See the discussion I had in the PR with Alex.  Basically, the purpose of this test is to make sure that we do the right thing when the service config fails validation, but the example here was being incorrectly failed by the old code.  So I changed the test to use an example that is actually invalid.

**Comment by markdroth** [comment]
Known issues: #21746 #21894 #21895 #13328 #21896

**Pull request merged**